src/hotspot/share/runtime/recursiveLock.cpp

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+#include "runtime/atomic.hpp"
+#include "runtime/recursiveLock.hpp"
+#include "runtime/thread.inline.hpp"
+
+RecursiveLock::RecursiveLock() : _sem(1), _owner(nullptr), _recursions(0) {}
+
+void RecursiveLock::lock(Thread* current) {
+  assert(current == Thread::current(), "must be current thread");
+  if (current == _owner) {
+    _recursions++;
+  } else {
+    // can be called by jvmti by VMThread.
+    if (current->is_Java_thread()) {
+      _sem.wait_with_safepoint_check(JavaThread::cast(current));
+    } else {
+      _sem.wait();
+    }
+    _recursions++;
+    assert(_recursions == 1, "should be");
+    Atomic::release_store(&_owner, current); // necessary?
```

Comment:
No release necessary. The only thing written since the sem.wait is recursions and it is only  updated or needed by the owning thread.

Reply:
ok thanks.